### PR TITLE
Allow cascading deletion of tree nodes and unused children

### LINKED ISSUE #1653: Unable to delete unused parent and children at once in the trees
**Describe the bug**
Trying to delete a parent and all its children that are all unused (no collection objects) is blocked

**To Reproduce**
1. Go to Ophidiidae genus in taxon tree
2. Click on 'edit' and then delete
3. See error

**Expected behavior**
You should be able to delete multiple nodes at once as long as the node and all of its children are unused

**Screenshots**
Capture

**Desktop (please complete the following information):**
 - OS: [e.g. Windows 10]
 - Browser [e.g. Chrome, Safari]
 - Specify 7 Version [e.g. 7.6.1]

**Reported By**
Name of your institution

**Additional context**

### PR DESCRIPTION
**Pull request by grantfitzsimmons**
Fixes #1653 

Rather than protecting children, the delete will cascade if there are no references to those children. 

If all child nodes are not referenced by any records within that discipline (taxon, geography, lithostrat, chronostrat, tectonicunit) or institution (storage), then the user should be able to delete the parent and all children:

https://github.com/user-attachments/assets/1b8054b2-db58-4945-a27f-f1eb15b69475

If any child is referenced by a record within that discipline or institution, it should block deletion:

https://github.com/user-attachments/assets/ebdb649a-b57d-4596-89be-c7b094e1bfa2

A new migration is included as well and it is reversible.

### Testing instructions

This PR affects database migrations. See migration testing instructions:
https://specify.github.io/testing/pull_requests.html#prs-tagged-with-label-migration

**Tip:** To find unused nodes, make the counts climb up the entire tree: https://discourse.specifysoftware.org/t/change-tree-node-count-behavior/2014/1

- [ ] If all child nodes are not referenced by any records within that discipline (taxon, geography, lithostrat, chronostrat, tectonicunit) or institution (storage), then the user should be able to delete the parent and all children.
- [ ] If any child nodes are used, deletion should be blocked.

### CODE REVIEW

File: specifyweb/specify/migrations/0010_cascading_tree_nodes.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
+# Generated by Django 3.2.15 on 2024-10-31 02:34
+
+from django.db import migrations, models
+import django.db.models.deletion
+
+
+class Migration(migrations.Migration):
+
+    dependencies = [
```
Comment (CarolineDenis) [inline specifyweb/specify/migrations/0010_cascading_tree_nodes.py:9]: TODO: 
Maybe change dependency once ready to merge

**Review by CarolineDenis — APPROVED** [review]

File: specifyweb/specify/migrations/0010_cascading_tree_nodes.py (hunk 1 above)
Comment (grantfitzsimmons) [reply specifyweb/specify/migrations/0010_cascading_tree_nodes.py:9]: 💯

**Review by sharadsw — APPROVED** [review]

**Review by alesan99 — APPROVED** [review]
- [x] If all child nodes are not referenced by any records within that discipline (taxon, geography, lithostrat, chronostrat, tectonicunit) or institution (storage), then the user should be able to delete the parent and all children.
- [x] If any child nodes are used, deletion should be blocked.

I am able to delete parents in the Taxon, Geography, and Storage trees 👍👍.
Everything with referenced children I've attempted to delete has successfully been blocked.

Unused synonyms also block deletion, which makes sense logically but I wanted to note it.

**Review by emenslin — APPROVED** [review]
### Testing instructions

This PR affects database migrations. See migration testing instructions:
https://specify.github.io/testing/pull_requests.html#prs-tagged-with-label-migration

**Tip:** To find unused nodes, make the counts climb up the entire tree: https://discourse.specifysoftware.org/t/change-tree-node-count-behavior/2014/1

- [x] If all child nodes are not referenced by any records within that discipline (taxon, geography, lithostrat, chronostrat, tectonicunit) or institution (storage), then the user should be able to delete the parent and all children.
- [x] If any child nodes are used, deletion should be blocked.

Worked in taxon, geography, and storage trees!

**Review by lexiclevenger — APPROVED** [review]
### Testing instructions

This PR affects database migrations. See migration testing instructions:
https://specify.github.io/testing/pull_requests.html#prs-tagged-with-label-migration

**Tip:** To find unused nodes, make the counts climb up the entire tree: https://discourse.specifysoftware.org/t/change-tree-node-count-behavior/2014/1

- [x] If all child nodes are not referenced by any records within that discipline (taxon, geography, lithostrat, chronostrat, tectonicunit) or institution (storage), then the user should be able to delete the parent and all children.
- [x] If any child nodes are used, deletion should be blocked.
---

Works for storage, taxon, geography, and stratigraphy trees!

**Pull request merged**